### PR TITLE
[llvm-dlltool] Handle MIPS R4000 architecture

### PR DESCRIPTION
**Pull request by hpoussin**

### CODE REVIEW

**Comment by hpoussin** [comment]
This is an extract of PR https://github.com/llvm/llvm-project/pull/107744

File: llvm/lib/ToolDrivers/llvm-dlltool/DlltoolDriver.cpp [hunk 1]
```diff
@@ -79,6 +79,7 @@ MachineTypes getEmulation(StringRef S) {
       .Case("arm", IMAGE_FILE_MACHINE_ARMNT)
       .Case("arm64", IMAGE_FILE_MACHINE_ARM64)
       .Case("arm64ec", IMAGE_FILE_MACHINE_ARM64EC)
+      .Case("mips", IMAGE_FILE_MACHINE_R4000)
```
Comment (compnerd) [inline llvm/lib/ToolDrivers/llvm-dlltool/DlltoolDriver.cpp:82]: Hmm, we should cross-check this with binutils - if it should be r4000 or mips as there was a MIPS16 and a MIPSv2 machine type that PE/COFF has supported over the years.

**Comment by hpoussin** [comment]
binutils `dlltool` only supports `i386`, `i386:x86-64`, `arm`, `thumb`, `arm64`, `arm-wince`, `arm_internetwork`, `mcore-be`, `mcore-le`, `mcore-elf-be` and `mcore-elf-le`.
It also supported once `ppc` (removed), `arm-epoc` (removed), `mcore` (renamed to `mcore-be`) and `mcore-elf` (renamed to `mcore-elf-be`)

`llvm-binutils` currently supports `i386`, `i386:x86-64`, `arm`, `arm64` and `arm64ec`.

`mips` name may fit in this list, with `mips16` and `mipsv2` later, or `mips-wince` for the Windows CE variant.

**Comment by compnerd** [comment]
What do you think of `-m r4000` as the parameter instead of `-m mips`?

**Comment by hpoussin** [comment]
I don't care.

**Comment by wzssyqa** [comment]
> What do you think of `-m r4000` as the parameter instead of `-m mips`?

Why r4000 is better than mips here?

**Comment by wzssyqa** [comment]
Oh, I see: https://learn.microsoft.com/en-us/windows/win32/sysinfo/image-file-machine-constants

```
IMAGE_FILE_MACHINE_R3000
0x0162
MIPS little-endian, 0x160 big-endian
IMAGE_FILE_MACHINE_R4000
0x0166
MIPS little-endian
IMAGE_FILE_MACHINE_R10000
0x0168
MIPS little-endian
IMAGE_FILE_MACHINE_WCEMIPSV2
0x0169
MIPS little-endian WCE v2
```

**Pull request merged**